### DISCUSSION TITLE
Install Additional Software

### DISCUSSION THREAD

**Discussion by stive101** [Ym0T/pterodactyl-nginx-egg#7] [General]
Hello,

I'm not a professional in this area, so I apologize for this question. I need Perl and additional modules to run a statistics website for my game server. However, there is only PHP available.

So, the only solution is to create my own Dockerfile and set it up there, right?

To try, I simply copied your Dockerfile and performed a Docker build, adding the new Docker image to the egg and starting the server with it.

However, what I'm encountering are some permission errors. I don't know why; shouldn't it be the same when pulling it from Git or using it locally?

thats the error i got : 

[Pterodactyl Daemon]: Finished pulling Docker container image
/docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
/docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
/docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
10-listen-on-ipv6-by-default.sh: info: can not modify /etc/nginx/conf.d/default.conf (read-only file system?)
/docker-entrypoint.sh: Sourcing /docker-entrypoint.d/15-local-resolvers.envsh
/docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh
/docker-entrypoint.sh: Launching /docker-entrypoint.d/30-tune-worker-processes.sh
/docker-entrypoint.sh: Configuration complete; ready for start up
2024/04/04 21:11:07 [warn] 1#1: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /etc/nginx/nginx.conf:2
nginx: [warn] the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /etc/nginx/nginx.conf:2
2024/04/04 21:11:07 [emerg] 1#1: mkdir() "/var/cache/nginx/client_temp" failed (30: Read-only file system)
nginx: [emerg] mkdir() "/var/cache/nginx/client_temp" failed (30: Read-only file system)

**Comment by Ym0T** [comment]
Hello,

can you send in your Dockerfile? Then we could have a look at it.
Have any other files been changed?

Regards

**Comment by stive101** [comment]
Hi,

Thank you for your prompt response. As I mentioned in the first post, I'm not changing the Docker file atm. I'm simply building it locally and adding it to the egg.

FROM debian:bookworm-slim

LABEL author="Ym0t" maintainer="YmoT@tuta.com"

ARG PHP_VERSION="8.3"

ENV DEBIAN_FRONTEND noninteractive

RUN apt-get update \
    && apt-get upgrade -y \
    && apt-get install -y apt-transport-https lsb-release ca-certificates wget nginx \
    && wget -O /etc/apt/trusted.gpg.d/php.gpg https://packages.sury.org/php/apt.gpg \
    && echo "deb https://packages.sury.org/php/ $(lsb_release -sc) main" | tee /etc/apt/sources.list.d/php.list \
    && apt-get update \
    && apt-get install -y --no-install-recommends \
        php${PHP_VERSION} \
        php${PHP_VERSION}-fpm \
        php${PHP_VERSION}-cli \
        php${PHP_VERSION}-common \
        php${PHP_VERSION}-mysqlnd \
        php${PHP_VERSION}-PDO \
        php${PHP_VERSION}-psr \
        php${PHP_VERSION}-xml \
        php${PHP_VERSION}-bcmath \
        php${PHP_VERSION}-calendar \
        php${PHP_VERSION}-ctype \
        php${PHP_VERSION}-curl \
        php${PHP_VERSION}-dom \
        php${PHP_VERSION}-mbstring \
        php${PHP_VERSION}-fileinfo \
        php${PHP_VERSION}-ftp \
        php${PHP_VERSION}-gd \
        php${PHP_VERSION}-gettext \
        php${PHP_VERSION}-gmp \
        php${PHP_VERSION}-iconv \
        php${PHP_VERSION}-igbinary \
        php${PHP_VERSION}-imagick \
        php${PHP_VERSION}-imap \
        php${PHP_VERSION}-intl \
        php${PHP_VERSION}-ldap \
        php${PHP_VERSION}-exif \
        php${PHP_VERSION}-memcache \
        php${PHP_VERSION}-mongodb \
        php${PHP_VERSION}-msgpack \
        php${PHP_VERSION}-mysqli \
        php${PHP_VERSION}-odbc \
        php${PHP_VERSION}-pcov \
        php${PHP_VERSION}-pgsql \
        php${PHP_VERSION}-Phar \
        php${PHP_VERSION}-posix \
        php${PHP_VERSION}-ps \
        php${PHP_VERSION}-pspell \
        php${PHP_VERSION}-readline \
        php${PHP_VERSION}-shmop \
        php${PHP_VERSION}-SimpleXML \
        php${PHP_VERSION}-soap \
        php${PHP_VERSION}-sockets \
        php${PHP_VERSION}-sqlite3 \
        php${PHP_VERSION}-sysvmsg \
        php${PHP_VERSION}-sysvsem \
        php${PHP_VERSION}-sysvshm \
        php${PHP_VERSION}-tokenizer \
        php${PHP_VERSION}-xmlreader \
        php${PHP_VERSION}-xmlwriter \
        php${PHP_VERSION}-xsl \
        php${PHP_VERSION}-zip \
        php${PHP_VERSION}-mailparse \
        php${PHP_VERSION}-memcached \
        php${PHP_VERSION}-inotify \
        php${PHP_VERSION}-maxminddb \
        php${PHP_VERSION}-protobuf \
        php${PHP_VERSION}-OPcache \
    && apt-get purge -y --auto-remove \
    && rm -rf /var/lib/apt/lists/*

RUN useradd -m -d /home/container/ -s /bin/bash container
ENV USER=container HOME=/home/container

WORKDIR /home/container

STOPSIGNAL SIGINT

COPY ./entrypoint.sh /entrypoint.sh
RUN chmod +x /entrypoint.sh

CMD /entrypoint.sh

**Comment by stive101** [comment]
im just need perl, maybe u can have a other solution then modifying the Dockerfile ?

**Reply by Ym0T** [reply]
Unfortunately, I have hardly any experience with perl. The first thing that comes to my mind is to simply modify the Dockerfile according to your wishes.

**Comment by Ym0T** [comment]
Sorry, I took that the wrong way. Define this in the Dockerfile, in which case you need to add this.
On the subject of permission errors. I don't know what your local configuration looks like, i.e. what your setup is.
Php and Nginx are specifically assigned the configuration file. In this case specially modified for the Docker container.

with Nginx, the startup command looks like this: `nginx -c /home/container/nginx/nginx.conf -p /home/container`

and for PHP:
`php-fpm$PHP_VERSION -c /home/container/php/php.ini --fpm-config /home/container/php/php-fpm.conf --daemonise > /dev/null 2>&1
`
In this case, the directory in which the configuration file and other web server files are found is specified. This can of course be different for a local system. It is actually best to test this correctly on a system that has already been set up.

I also had the permissions error, or at least a similar one, when creating this egg, but this was due to the configuration file as described.

**Reply by stive101** [reply]
hi, 

thank you for your replay,
i got it to work now, 
i was wondering how big the Perl modules ar 1.5 gb in total o.o. 

as the compiling starts,i saw, took hours to compile XD. But finaly should work

thank you.

**Reply by Ym0T** [reply]
Have you now implemented it with a Dockerdfile and what was the problem? If someone should have the same problem. :)

**Reply by stive101** [reply]
the error above, im not realy sure how i solved it, idk why but the docker file didnt compile with the correct entrypoint. i deleting all images i dont use anymore, also the containers and try to recomile the image file and then it workd. 

for Perl Modules, im not atm sure if the way i did is correct, need to test some stuff, maybe i get my perl scripts to work with apt packages, that already exists so i dont neet to compile that mutch. but maybe the modules i try to install needs all that dependancys, 

i will see, have to test that with apt packages

**Reply by Ym0T** [reply]
All right, thanks for the answer!

**Reply by stive101** [reply]
thanks for your help :P
